File: benchmarks/response_overhead.go
// Microbenchmark: response parsing overhead on the Go side.
//
// Measures the cost of each step in the ack response path:
// 1. String comparison (current fast path)
// 2. json.Unmarshal (slow path / baseline)
// 3. Hypothetical binary: read status byte
//
// Run: cd benchmarks && go test -bench=. -benchmem response_overhead_test.go

package main

import (
	"encoding/json"
	"fmt"
	"runtime"
	"time"
	"unsafe"
)

const (
	iterations = 5_000_000
	warmup     = 500_000

	ackCompact = `{"status":"ack"}`
	ackSpaced  = `{"status": "ack"}`

	violationJSON = `{"type":"property","status":"fails","property_index":0,` +
		`"timestamp":1234567890,"reason":"Signal EngineSpeed exceeded 220"}`
)

func bench(name string, f func(), iters int) float64 {
	// Warmup
	for i := 0; i < warmup; i++ {
		f()
	}
	runtime.GC()

	start := time.Now()
	for i := 0; i < iters; i++ {
		f()
	}
	elapsed := time.Since(start)
	nsPerOp := float64(elapsed.Nanoseconds()) / float64(iters)
	fmt.Printf("  %-45s  %8.1f ns/op\n", name, nsPerOp)
	return nsPerOp
}

// noescape prevents the compiler from optimizing away allocations.
//
//go:noinline
func sink(v any) { _ = v }

func main() {
	fmt.Println("Response parsing overhead (Go)")
	fmt.Printf("Iterations: %d\n\n", iterations)

	ackStr := ackCompact
	violStr := violationJSON

	fmt.Println("=== Ack response (hot path, ~99% of frames) ===")

	// 1. String comparison (current fast path)
	tCmp := bench("string comparison", func() {
		is := (ackStr == ackCompact) || (ackStr == ackSpaced)
		sink(is)
	}, iterations)

	// 2. json.Unmarshal to map
	tJSON := bench("json.Unmarshal(ack)", func() {
		var m map[string]any
		_ = json.Unmarshal([]byte(ackStr), &m)
		sink(m)
	}, iterations)

	// 3. json.Unmarshal to struct
	type ackResp struct {
		Status string `json:"status"`
	}
	tJSONStruct := bench("json.Unmarshal(ack, struct)", func() {
		var r ackResp
		_ = json.Unmarshal([]byte(ackStr), &r)
		sink(r)
	}, iterations)

	// 4. Hypothetical binary: read status byte
	binaryAck := []byte{0x00}
	tBin := bench("binary: read status byte", func() {
		is := binaryAck[0] == 0x00
		sink(is)
	}, iterations)

	// 5. Simulated CString → Go string (unsafe, like cgo does)
	cBytes := []byte(ackCompact + "\x00")
	cPtr := unsafe.Pointer(&cBytes[0])
	tCgo := bench("C.GoString simulation (16 bytes)", func() {
		// Simulate what cgo's C.GoString does: find null, copy bytes
		p := (*[64]byte)(cPtr)
		n := 0
		for p[n] != 0 {
			n++
		}
		s := string(p[:n])
		sink(s)
	}, iterations)

	fmt.Println()
	budget := 1_000_000_000.0 / 48_000 // ~20.8 us at 48k fps
	fmt.Printf("  Fast path saves vs json.Unmarshal: %8.1f ns/frame\n", tJSON-tCmp)
	fmt.Printf("  Binary saves vs fast path:         %8.1f ns/frame\n", tCmp-tBin)
	fmt.Printf("  CString→Go string overhead:        %8.1f ns/frame\n", tCgo)
	fmt.Printf("\n=== Context ===\n")
	fmt.Printf("  Per-frame budget at 48k fps:       %8.0f ns\n", budget)
	fmt.Printf("  Fast path %% of budget:             %8.1f%%\n", tCmp/budget*100)
	fmt.Printf("  json.Unmarshal %% of budget:        %8.1f%%\n", tJSON/budget*100)
	fmt.Printf("  json struct %% of budget:           %8.1f%%\n", tJSONStruct/budget*100)
	fmt.Printf("  Binary %% of budget:                %8.1f%%\n", tBin/budget*100)

	// --- Violation path ---
	fmt.Println("\n=== Violation response (rare) ===")

	bench("json.Unmarshal(violation, map)", func() {
		var m map[string]any
		_ = json.Unmarshal([]byte(violStr), &m)
		sink(m)
	}, 1_000_000)

	bench("[]byte conversion (violation)", func() {
		b := []byte(violStr)
		sink(b)
	}, 1_000_000)
}

File: benchmarks/response_overhead.cpp
// Microbenchmark: response parsing overhead on the C++ side.
//
// Measures the cost of each step in the ack response path:
// 1. string_view comparison (current fast path)
// 2. nlohmann::json::parse (slow path / baseline)
// 3. Hypothetical binary: read status byte
//
// Build: g++-15 -std=c++23 -O3 -o response_overhead benchmarks/response_overhead.cpp
// Run:   ./response_overhead

#include <chrono>
#include <cstdint>
#include <cstdio>
#include <cstring>
#include <string>
#include <string_view>
#include <variant>

// Minimal JSON parsing simulation (no nlohmann dependency for standalone build)
// We'll time string operations and simulate the parse cost.

static constexpr std::string_view ack_compact = R"({"status":"ack"})";
static constexpr std::string_view ack_spaced  = R"({"status": "ack"})";

static constexpr std::string_view violation_json =
    R"({"type":"property","status":"fails","property_index":0,)"
    R"("timestamp":1234567890,"reason":"Signal EngineSpeed exceeded 220"})";

struct Ack {};
struct Violation { int property_index; uint64_t timestamp; std::string reason; };
using FrameResponse = std::variant<Ack, Violation>;

static constexpr int ITERATIONS = 5'000'000;
static constexpr int WARMUP     = 500'000;

template <typename F>
auto bench(const char* name, F&& func, int iterations = ITERATIONS) -> double {
    // Warmup
    for (int i = 0; i < WARMUP; ++i)
        func();

    auto start = std::chrono::high_resolution_clock::now();
    for (int i = 0; i < iterations; ++i)
        func();
    auto end = std::chrono::high_resolution_clock::now();

    double ns = std::chrono::duration<double, std::nano>(end - start).count();
    double ns_per_op = ns / iterations;
    std::printf("  %-45s  %8.1f ns/op\n", name, ns_per_op);
    return ns_per_op;
}

// Prevent optimizing away results
template <typename T>
void do_not_optimize(T const& val) {
    asm volatile("" : : "r,m"(val) : "memory");
}

int main() {
    std::puts("Response parsing overhead (C++)");
    std::printf("Iterations: %'d\n\n", ITERATIONS);

    // Build test strings (simulate what FFI returns)
    std::string ack_str{ack_compact};
    std::string viol_str{violation_json};

    std::puts("=== Ack response (hot path, ~99% of frames) ===");

    // 1. string_view comparison (current fast path)
    auto t_cmp = bench("string_view comparison", [&] {
        std::string_view sv{ack_str};
        bool is_ack = (sv == ack_compact) || (sv == ack_spaced);
        do_not_optimize(is_ack);
    });

    // 2. std::string construction from C string (simulates CString → string)
    const char* c_ack = ack_str.c_str();
    auto t_str = bench("std::string from c_str", [&] {
        std::string s{c_ack};
        do_not_optimize(s);
    });

    // 3. Full fast path: string construction + comparison
    auto t_fast = bench("string + comparison (full fast)", [&] {
        std::string s{c_ack};
        std::string_view sv{s};
        bool is_ack = (sv == ack_compact) || (sv == ack_spaced);
        do_not_optimize(is_ack);
    });

    // 4. String scan for "ack" (simulates minimal JSON field check)
    auto t_find = bench("string::find(\"ack\")", [&] {
        std::string s{c_ack};
        bool found = s.find("ack") != std::string::npos;
        do_not_optimize(found);
    });

    // 5. strcmp (raw C comparison, no string construction)
    auto t_strcmp = bench("strcmp (raw C)", [&] {
        bool is_ack = (std::strcmp(c_ack, "{\"status\":\"ack\"}") == 0);
        do_not_optimize(is_ack);
    });

    // 6. Hypothetical binary: read status byte
    uint8_t binary_ack = 0x00;
    auto t_bin = bench("binary: read status byte", [&] {
        bool is_ack = (binary_ack == 0x00);
        do_not_optimize(is_ack);
    });

    // 7. memcmp (fixed-size comparison)
    auto t_memcmp = bench("memcmp (16 bytes)", [&] {
        bool is_ack = (std::memcmp(c_ack, "{\"status\":\"ack\"}", 16) == 0);
        do_not_optimize(is_ack);
    });

    std::puts("");
    double budget = 1'000'000'000.0 / 48'000;  // ~20.8 us at 48k fps
    std::printf("  Fast path saves vs string+cmp:  %8.1f ns/frame\n", t_str - t_cmp);
    std::printf("  Binary saves vs fast path:      %8.1f ns/frame\n", t_fast - t_bin);
    std::printf("\n=== Context ===\n");
    std::printf("  Per-frame budget at 48k fps:    %8.0f ns\n", budget);
    std::printf("  Fast path %% of budget:          %8.1f%%\n", t_fast / budget * 100);
    std::printf("  Binary %% of budget:             %8.1f%%\n", t_bin / budget * 100);

    // --- Violation path (for reference) ---
    std::puts("\n=== Violation response (rare) ===");

    const char* c_viol = viol_str.c_str();
    bench("std::string from c_str (violation)", [&] {
        std::string s{c_viol};
        do_not_optimize(s);
    }, 2'000'000);

    // Manual field extraction (simulates what we'd do without nlohmann)
    bench("find fields manually", [&] {
        std::string s{c_viol};
        auto idx = s.find("\"property_index\":");
        auto ts = s.find("\"timestamp\":");
        do_not_optimize(idx);
        do_not_optimize(ts);
    }, 2'000'000);

    return 0;
}
